File: SU2_CFD/src/SU2_CFD.cpp
/*!
 * \file SU2_CFD.cpp
 * \brief Main file of the SU2 Computational Fluid Dynamics code
 * \author F. Palacios, T. Economon
 * \version 8.0.0 "Harrier"
 *
 * SU2 Project Website: https://su2code.github.io
 *
 * The SU2 Project is maintained by the SU2 Foundation
 * (http://su2foundation.org)
 *
 * Copyright 2012-2023, SU2 Contributors (cf. AUTHORS.md)
 *
 * SU2 is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * SU2 is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with SU2. If not, see <http://www.gnu.org/licenses/>.
 */

#include "../include/SU2_CFD.hpp"

/* LIBXSMM include files, if supported. */
#ifdef HAVE_LIBXSMM
#include "libxsmm.h"
#endif

/* Include file, needed for the runtime NaN catching. You also have to include feenableexcept(...) below. */
//#include <fenv.h>

using namespace std;

int main(int argc, char *argv[]) {

  char config_file_name[MAX_STRING_SIZE];
  bool dry_run = false;
  int num_threads = omp_get_max_threads();
  bool use_thread_mult = false;
  std::string filename = "default.cfg";

  /*--- Command line parsing ---*/

  CLI::App app{"SU2 v8.0.0 \"Harrier\", The Open-Source CFD Code"};
  app.add_flag("-d,--dryrun", dry_run, "Enable dry run mode.\n"
                                       "Only execute preprocessing steps using a dummy geometry.");
  app.add_option("-t,--threads", num_threads, "Number of OpenMP threads per MPI rank.");
  app.add_flag("--thread_multiple", use_thread_mult, "Request MPI_THREAD_MULTIPLE thread support.");
  app.add_option("configfile", filename, "A config file.")->check(CLI::ExistingFile);

  CLI11_PARSE(app, argc, argv)

  /*--- OpenMP setup ---*/

  omp_set_num_threads(num_threads);

  /*--- MPI initialization, and buffer setting ---*/

#if defined(HAVE_OMP) && defined(HAVE_MPI)
  int required = use_thread_mult? MPI_THREAD_MULTIPLE : MPI_THREAD_FUNNELED;
  int provided;
  SU2_MPI::Init_thread(&argc, &argv, required, &provided);
#else
  SU2_MPI::Init(&argc, &argv);
#endif
  SU2_MPI::Comm MPICommunicator = SU2_MPI::GetComm();

  /*--- Further initializations are placed in the constructor of CDriverBase, to ensure that they are also seen by the
   python wrapper. */

  /*--- Uncomment the following line if runtime NaN catching is desired. ---*/
  // feenableexcept(FE_INVALID | FE_OVERFLOW | FE_DIVBYZERO );

  /*--- Initialize libxsmm, if supported. ---*/
#ifdef HAVE_LIBXSMM
  libxsmm_init();
#endif

  /*--- Create a pointer to the main SU2 Driver ---*/

  CDriver* driver = nullptr;

  /*--- Load in the number of zones and spatial dimensions in the mesh file (If no config
   file is specified, default.cfg is used) ---*/
  strcpy(config_file_name, filename.c_str());

  /*--- Read the name and format of the input mesh file to get from the mesh
   file the number of zones and dimensions from the numerical grid (required
   for variables allocation). ---*/

  const CConfig config(config_file_name, SU2_COMPONENT::SU2_CFD);
  const unsigned short nZone = config.GetnZone();

  /*--- First, given the basic information about the number of zones and the
   solver types from the config, instantiate the appropriate driver for the problem
   and perform all the preprocessing. ---*/

  const bool disc_adj = config.GetDiscrete_Adjoint();
  const bool multizone = config.GetMultizone_Problem();
  const bool harmonic_balance = (config.GetTime_Marching() == TIME_MARCHING::HARMONIC_BALANCE);

  if (dry_run) {

    /*--- Dry Run. ---*/
    driver = new CDummyDriver(config_file_name, nZone, MPICommunicator);

  }
  else if (!multizone && !harmonic_balance) {

    /*--- Generic single zone problem: instantiate the single zone driver class. ---*/
    if (nZone != 1)
      SU2_MPI::Error("The required solver doesn't support multizone simulations", CURRENT_FUNCTION);

    if (disc_adj) {
      driver = new CDiscAdjSinglezoneDriver(config_file_name, nZone, MPICommunicator);
    }
    else {
      driver = new CSinglezoneDriver(config_file_name, nZone, MPICommunicator);
    }

  }
  else if (multizone) {

    /*--- Generic multizone problems. ---*/
    if (disc_adj) {
      driver = new CDiscAdjMultizoneDriver(config_file_name, nZone, MPICommunicator);
    }
    else {
      driver = new CMultizoneDriver(config_file_name, nZone, MPICommunicator);
    }

  }
  else {
    assert(harmonic_balance);

    /*--- Harmonic balance problem: instantiate the Harmonic Balance driver class. ---*/
    driver = new CHBDriver(config_file_name, nZone, MPICommunicator);

  }

  /*--- Launch the main external loop of the solver. ---*/

  driver->StartSolver();

  /*--- Finalize solver, delete all the containers, close history file, exit SU2. ---*/

  driver->Finalize();

  delete driver;

  /*---Finalize libxsmm, if supported. ---*/
#ifdef HAVE_LIBXSMM
  libxsmm_finalize();
#endif

  /*--- Finalize AD. ---*/
  AD::Finalize();

  /*--- Finalize MPI parallelization. ---*/
  SU2_MPI::Finalize();

  /*--- Finalize OpenMP. ---*/
  omp_finalize();

  return EXIT_SUCCESS;

}

File: SU2_PY/SU2_CFD.py
#!/usr/bin/env python

## \file SU2_CFD.py
#  \brief Python script to launch SU2_CFD through the Python Wrapper.
#  \author David Thomas
#  \version 8.0.0 "Harrier"
#
# SU2 Project Website: https://su2code.github.io
#
# The SU2 Project is maintained by the SU2 Foundation
# (http://su2foundation.org)
#
# Copyright 2012-2023, SU2 Contributors (cf. AUTHORS.md)
#
# SU2 is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# SU2 is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with SU2. If not, see <http://www.gnu.org/licenses/>.

# ----------------------------------------------------------------------
#  Imports
# ----------------------------------------------------------------------

from __future__ import division, print_function, absolute_import
from optparse import OptionParser  # use a parser for configuration
import SU2  # imports SU2 python tools
import pysu2  # imports the SU2 wrapped module

# -------------------------------------------------------------------
#  Main
# -------------------------------------------------------------------


def main():

    # Command line options
    parser = OptionParser()
    parser.add_option(
        "-f", "--file", dest="filename", help="Read config from FILE", metavar="FILE"
    )
    parser.add_option(
        "--nDim",
        dest="nDim",
        default=2,
        help="Define the number of DIMENSIONS",
        metavar="DIMENSIONS",
    )
    parser.add_option(
        "--nZone",
        dest="nZone",
        default=1,
        help="Define the number of ZONES",
        metavar="NZONE",
    )
    parser.add_option(
        "--parallel",
        action="store_true",
        help="Specify if we need to initialize MPI",
        dest="with_MPI",
        default=False,
    )
    parser.add_option(
        "--fsi",
        dest="fsi",
        default="False",
        help="Launch the FSI driver",
        metavar="FSI",
    )
    parser.add_option(
        "--fem",
        dest="fem",
        default="False",
        help="Launch the FEM driver (General driver)",
        metavar="FEM",
    )
    parser.add_option(
        "--harmonic_balance",
        dest="harmonic_balance",
        default="False",
        help="Launch the Harmonic Balance (HB) driver",
        metavar="HB",
    )
    parser.add_option(
        "--poisson_equation",
        dest="poisson_equation",
        default="False",
        help="Launch the poisson equation driver (General driver)",
        metavar="POIS_EQ",
    )
    parser.add_option(
        "--wave_equation",
        dest="wave_equation",
        default="False",
        help="Launch the wave equation driver (General driver)",
        metavar="WAVE_EQ",
    )
    parser.add_option(
        "--heat_equation",
        dest="heat_equation",
        default="False",
        help="Launch the heat equation driver (General driver)",
        metavar="HEAT_EQ",
    )

    (options, args) = parser.parse_args()
    options.nDim = int(options.nDim)
    options.nZone = int(options.nZone)
    options.fsi = options.fsi.upper() == "TRUE"
    options.fem = options.fem.upper() == "TRUE"
    options.harmonic_balance = options.harmonic_balance.upper() == "TRUE"
    options.poisson_equation = options.poisson_equation.upper() == "TRUE"
    options.wave_equation = options.wave_equation.upper() == "TRUE"
    options.heat_equation = options.heat_equation.upper() == "TRUE"

    if options.filename == None:
        raise Exception("No config file provided. Use -f flag")

    if options.with_MPI == True:
        from mpi4py import MPI  # use mpi4py for parallel run (also valid for serial)

        comm = MPI.COMM_WORLD
    else:
        comm = 0

    # Initialize the corresponding driver of SU2, this includes solver preprocessing
    try:
        if (options.nZone == 1) and (
            options.fem
            or options.poisson_equation
            or options.wave_equation
            or options.heat_equation
        ):
            SU2Driver = pysu2.CSinglezoneDriver(options.filename, options.nZone, comm)
        elif options.harmonic_balance:
            SU2Driver = pysu2.CHBDriver(options.filename, options.nZone, comm)
        elif options.nZone >= 2:
            SU2Driver = pysu2.CMultizoneDriver(options.filename, options.nZone, comm)
        else:
            SU2Driver = pysu2.CSinglezoneDriver(options.filename, options.nZone, comm)
    except TypeError as exception:
        print("A TypeError occured in pysu2.CDriver : ", exception)
        if options.with_MPI == True:
            print(
                "ERROR : You are trying to initialize MPI with a serial build of the wrapper. Please, remove the --parallel option that is incompatible with a serial build."
            )
        else:
            print(
                "ERROR : You are trying to launch a computation without initializing MPI but the wrapper has been built in parallel. Please add the --parallel option in order to initialize MPI for the wrapper."
            )
        return

    # Launch the solver for the entire computation
    SU2Driver.StartSolver()

    # Finalize the solver and exit cleanly
    SU2Driver.Finalize()

    if SU2Driver != None:
        del SU2Driver


# -------------------------------------------------------------------
#  Run Main Program
# -------------------------------------------------------------------

# this is only accessed if running from command prompt
if __name__ == "__main__":
    main()
